ethsigner/app/src/main/java/tech/pegasys/ethsigner/EthSignerApp.java

```diff
@@ -12,31 +12,65 @@
  */
 package tech.pegasys.ethsigner;
 
-import tech.pegasys.ethsigner.core.RunnerBuilder;
+import tech.pegasys.ethsigner.core.signing.TransactionSigner;
+import tech.pegasys.ethsigner.core.signing.fileBased.FileBasedSignerBuilder;
+import tech.pegasys.ethsigner.core.signing.hashicorp.HashicorpSignerBuilder;
 
+import io.vertx.core.Vertx;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.core.config.Configurator;
 
 public class EthSignerApp {
 
   private static final Logger LOG = LogManager.getLogger();
+  private static final Vertx vertx = Vertx.vertx();
 
   public static void main(final String... args) {
     final CommandLineConfig config = new CommandLineConfig(System.out);
     if (!config.parse(args)) {
       return;
     }
 
+    Runtime.getRuntime().addShutdownHook(new Cleanup());
+
     // set log level per CLI flags
     System.out.println("Setting logging level to " + config.getLogLevel().name());
     Configurator.setAllLevels("", config.getLogLevel());
 
     LOG.debug("Configuration = {}", config);
     LOG.info("Version = {}, ", ApplicationInfo.version());
 
+    // create a signer based on the configuration provided
+    final TransactionSigner signer = createTransactionSigner(config);
+
+    if (signer == null) {
+      LOG.error("Cannot create a signer from the given config: " + config);
+      System.exit(-1);
+    }
+
     final tech.pegasys.ethsigner.core.EthSigner ethSigner =
-        new tech.pegasys.ethsigner.core.EthSigner(config, new RunnerBuilder());
+        new tech.pegasys.ethsigner.core.EthSigner(config, signer, vertx);
     ethSigner.run();
   }
+
+  private static TransactionSigner createTransactionSigner(final CommandLineConfig config) {
+    TransactionSigner signer = null;
+    if (config.getHashicorpSignerConfig().isConfigured()) {
+      signer = new HashicorpSignerBuilder(config.getHashicorpSignerConfig(), vertx).build();
+    } else if (config.getFileBasedSignerConfig().isConfigured()) {
+      signer = new FileBasedSignerBuilder(config.getFileBasedSignerConfig()).build();
+    }
+    return signer;
+  }
+
+  private static class Cleanup extends Thread {
+
+    @Override
+    public void run() {
+      if (vertx != null) {
+        vertx.close();
```

Comment:
Does this really need calling? Isn't Vertx a deamon thread?

Reply:
Closing vertx is a very good idea.  It gives everything deployed in it a chance to shutdown cleanly, including closing TCP sockets nicely so they don't wind up stuck in TCP_WAIT etc.

Reply:
Seems like not
>Threads maintained by Vert.x instances are not daemon threads so they will prevent the JVM from exiting.

> If you are embedding Vert.x and you have finished with it, you can call close to close it down.
https://vertx.io/docs/vertx-core/java/

Reply:
I'm not sure. I think it is always a good idea to call close on something so it releases it's resources.

Reply:
I assumed that the closes method is there for a reason ...